builder/dispatchers.go

```diff
@@ -226,10 +226,16 @@ func run(b *Builder, args []string, attributes map[string]bool, original string)
 	// set Cmd manually, this is special case only for Dockerfiles
 	b.Config.Cmd = config.Cmd
 	runconfig.Merge(b.Config, config)
+	// set build-time environment for 'run'. We let dockerfile
+	// environment override build-time environment.
+	env := b.Config.Env
```

Comment:
re: expansion - see https://github.com/docker/docker/blob/master/builder/support.go#L18 and then look for where that method is called.

To the other points, I need to think more about this because I may be incorrectly mixing topics and purposes.  The reason I thought of this was because I could image someone wanting their Dockerfile to have something like:
    COPY files-${arch}/Makefile /home/user/
where a different Makefile is copied based on the value of "arch" - meaning "x86" vs "arm". So, I thought it would be nice if they could run:
    docker build -e arch=arm .
But, my proposal sets 'arch' in the container too which is probably a bad thing based on the original issue/thread.

I may need to just "deal with it" but it feels odd to have env vars that only apply to certain Dockerfile commands and not all of them.  
I wonder if the notion of "build-time env vars" is something to consider.  Meaning, we create a sibling to b.Config.Env called b.Config.BuildEnv and those are available for variable expansion during the processing of the Dockerfile (and passed into the RUN command) but are NOT set within the container.

@erikh your thoughts?